### PR TITLE
Fix edge case for limb overflow in SMOD

### PR DESCRIPTION
**Pull request by daniellehrner**
## PR description

The optimized SMOD implementation does not handle one specific edge case correcty. This PR fixes that and adds a regression tests for this specific edge case

### Thanks for sending a pull request! Have you done the following?

- [ ] Checked out our [contribution guidelines](https://github.com/hyperledger/besu/blob/main/CONTRIBUTING.md)?
- [ ] Considered documentation and added the `doc-change-required` label to this PR [if updates are required](https://wiki.hyperledger.org/display/BESU/Documentation).
- [ ] Considered the changelog and included an [update if required](https://wiki.hyperledger.org/display/BESU/Changelog).
- [ ] For database changes (e.g. KeyValueSegmentIdentifier) considered compatibility and performed forwards and backwards compatibility tests

### Locally, you can run these tests to catch failures early:

- [ ] spotless: `./gradlew spotlessApply`
- [ ] unit tests: `./gradlew build`
- [ ] acceptance tests: `./gradlew acceptanceTest`
- [ ] integration tests: `./gradlew integrationTest`
- [ ] reference tests: `./gradlew ethereum:referenceTests:referenceTests`
- [ ] hive tests: [Engine or other RPCs modified?](https://lf-hyperledger.atlassian.net/wiki/spaces/BESU/pages/22156302/Using+Hive+Test+Suite)

### CODE REVIEW

File: evm/src/test/java/org/hyperledger/besu/evm/UInt256Test.java [hunk 1]
```diff
@@ -544,6 +544,46 @@ public void mulMod() {
     }
   }
 
+  @Test
+  public void modWithOverflowQuotientEstimate() {
```
Comment (lu-pinto) [inline evm/src/test/java/org/hyperledger/besu/evm/UInt256Test.java]: All 3 tests only target `UInt192::mulSubOverflow`. Please write up additional tests that can test the other 2 cases.

File: evm/src/main/java/org/hyperledger/besu/evm/UInt256.java [hunk 2]
```diff
@@ -1660,6 +1671,15 @@ private UInt256 mulSubOverflow(final long v3, final long v2, final long v1, fina
...
       long z3 = v3 + u3 - carry - borrow;
+      // q = MAX may still be 1 too high; check if result >= modulus (i.e. negative wrapped)
+      if (Long.compareUnsigned(z3, u3) > 0
+          || (z3 == u3
+              && (Long.compareUnsigned(z2, u2) > 0
+                  || (z2 == u2
+                      && (Long.compareUnsigned(z1, u1) > 0
+                          || (z1 == u1 && Long.compareUnsigned(z0, u0) >= 0)))))) {
+        return addBack(z3, z2, z1, z0);
+      }
```
Comment (lu-pinto) [inline evm/src/main/java/org/hyperledger/besu/evm/UInt256.java:1695]: I think it should be possible to write this without branches - did you check if this produces branching at assembly level?

Comment (lu-pinto) [reply evm/src/main/java/org/hyperledger/besu/evm/UInt256.java:1695]: Was thinking about comparing the last limb instead:
```
      long t1 = v3 + u3;
      long t2 = t1 - carry;
      long z3 = t2 - borrow;

      borrow =
          ((Long.compareUnsigned(t1, t2) < 0) ? 1L : 0L)
        + ((Long.compareUnsigned(t2, z3) < 0) ? 1L : 0L)
        - ((Long.compareUnsigned(t1, v3) < 0) ? 1L : 0L);
      if (borrow > 0) {
		return addBack(z3, z2, z1, z0);
      }
```
It passes all tests, but after consideration I think it's safer to go with your version as it seems safer to compare each limb one by one with the modulus.
Also from a performance perspective this is such a rare case that it doesn't really matter much.

Comment (thomas-quadratic) [reply evm/src/main/java/org/hyperledger/besu/evm/UInt256.java:1695]: Here a check that multipy substract with overflow gives a negative answer is sufficient, meaning there is a borrow, exactly like multiply substract operation.
I am not sure why you say it is safer, it is logically the same to check for a leftover borrow or to check the result has wrapped around, and it is similar to what we do in mulSub.

**Comment by thomas-quadratic** [comment]
Great, good catch for this bug that was introduced when we moved from div3by2 to div2by1. Quotient estimates can be too large up to 2 now, and this includes the overflow case.
I would approve this PR if I could.

File: evm/src/main/java/org/hyperledger/besu/evm/UInt256.java
Comment (lu-pinto) [inline evm/src/main/java/org/hyperledger/besu/evm/UInt256.java:1737]: don't think this needs a check `if (Long.compareUnsigned(v.u8, u3) >= 0)` as it looks like dead code

Comment (lu-pinto) [inline evm/src/main/java/org/hyperledger/besu/evm/UInt256.java:1509]: don't think this needs a check `if (Long.compareUnsigned(v.u7, u2) >= 0)` as it looks like dead code

**Comment by lu-pinto** [comment]
Merging it since @thomas-quadratic agreed to merge this
